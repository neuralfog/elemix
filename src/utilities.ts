import { activeRenderers } from './renderers';

export const ref = <Value>(
    value: Value | undefined = undefined,
): { value?: Value } => {
    return {
        value,
    };
};

export const fastUID = (): string =>
    Math.floor(performance.now() * 1000).toString(36) +
    Math.random().toString(36).slice(2, 6);

export function render(): Promise<boolean> {
    return new Promise((resolve) => {
        const check = () => {
            if (activeRenderers.size === 0) {
                resolve(false);
            } else {
                setTimeout(check, 0);
            }
        };
        check();
    });
}

export const camelToKebabCase = (input: string): string => {
    return (
        input.match(
            /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|[0-9]*$)|[A-Z]?[a-z]+|[A-Z]|[0-9]+/g,
        ) || []
    )
        .map((x) => x.toLowerCase())
        .join('-');
};
