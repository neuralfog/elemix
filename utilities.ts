export { mergeClasses } from '@neuralfog/elemix-renderer/utils';
export * from './src/utilities';
