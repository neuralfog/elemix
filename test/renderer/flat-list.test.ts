import { expect, test, describe, beforeEach } from 'vitest';
import { html } from '@neuralfog/elemix-renderer';
import { Present } from '@neuralfog/elemix-testing';
import { render } from '../../utilities';
import type { RendererList } from '../fixtures/renderer/RendererList';
import { HTML } from '@neuralfog/elemix-testing/snapshots';

import '../fixtures/renderer/RendererList';

describe('Renderer Flat List', () => {
    beforeEach(() => {
        document.body.innerHTML = '';
    });

    test('render', async () => {
        // Initial State
        const presenter = new Present().screen(html`<renderer-list />`);
        const component = presenter.getComponent<RendererList>('renderer-list');
        await render();
        expect(HTML(presenter.root<RendererList>())).toMatchSnapshot();

        // Move elements and add new
        component.state.list = [
            'Sophia Brown',
            'Olivia Davis',
            'John Doe',
            'Benjamin Williams',
            'Emily Johnson',
            'Michael Smith',
        ];
        await render();
        expect(HTML(presenter.root<RendererList>())).toMatchSnapshot();

        // Remove Elements
        component.state.list = ['Michael Smith', 'Benjamin Williams'];
        await render();
        expect(HTML(presenter.root<RendererList>())).toMatchSnapshot();

        // Empty List
        component.state.list = [];
        await render();
        expect(HTML(presenter.root<RendererList>())).toMatchSnapshot();
    });
});
