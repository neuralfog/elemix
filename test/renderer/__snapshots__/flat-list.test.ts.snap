// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renderer Flat List > render 1`] = `
"
<renderer-list><!-- #shadowroot -->
  <ul>
    <li>Emily Johnson<!--₥0--></li>
    <li>Michael Smith<!--₥0--></li>
    <li>Olivia Davis<!--₥0--></li>
    <li>Benjamin Williams<!--₥0--></li>
    <li>Sophia Brown<!--₥0--></li>
  </ul>
</renderer-list>
"
`;

exports[`Renderer Flat List > render 2`] = `
"
<renderer-list><!-- #shadowroot -->
  <ul>
    <li>Sophia Brown<!--₥0--></li>
    <li>Olivia Davis<!--₥0--></li>
    <li>John Doe<!--₥0--></li>
    <li>Benjamin Williams<!--₥0--></li>
    <li>Emily Johnson<!--₥0--></li>
    <li>Michael Smith<!--₥0--></li>
  </ul>
</renderer-list>
"
`;

exports[`Renderer Flat List > render 3`] = `
"
<renderer-list><!-- #shadowroot -->
  <ul>
    <li>Michael Smith<!--₥0--></li>
    <li>Benjamin Williams<!--₥0--></li>
  </ul>
</renderer-list>
"
`;

exports[`Renderer Flat List > render 4`] = `
"
<renderer-list><!-- #shadowroot -->
  <ul><!--₥0--></ul>
</renderer-list>
"
`;
